OpenSim/Common/DelimFileAdapter.h

```diff
@@ -60,17 +115,650 @@ class OSIMCOMMON_API DelimFileAdapter : public FileAdapter {
+    for(const auto& key : table->getTableMetaDataKeys()) {
+        try {
+            if(key != "header")
+                out_stream << key << "=" 
+                           << table->
+                              template getTableMetaData<std::string>(key) 
+                           << "\n";
+        } catch(const InvalidTemplateArgument&) {}
+    }
+    // Write name of the data-type -- vec3, vec6, etc.
+    out_stream << _dataTypeString << "=" << dataTypeName() << "\n";
+    // Write version number.
+    out_stream << _versionString << "=" << _versionNumber << "\n";
+    out_stream << _endHeaderString << "\n";
+
+    // Line containing column labels.
+    out_stream << _timeColumnLabel;
+    for(unsigned col = 0; col < table->getNumColumns(); ++col)
+        out_stream << _delimiterWrite
+                   << table->
+                      getDependentsMetaData().
+                      getValueArrayForKey("labels")[col].
+                      template getValue<std::string>();
+    out_stream << "\n";
+
+    // Data rows.
+    for(unsigned row = 0; row < table->getNumRows(); ++row) {
+        constexpr auto prec = std::numeric_limits<double>::digits10 + 1;
+        out_stream << std::setprecision(prec)
+                   << table->getIndependentColumn()[row];
+        const auto& row_r = table->getRowAtIndex(row);
+        for(unsigned col = 0; col < table->getNumColumns(); ++col) {
+            const auto& elt = row_r[col];
+            out_stream << _delimiterWrite;
+            writeElem(out_stream, elt, prec);
+        }
+        out_stream << "\n";
+    }
+}
+
+template<>
+inline void 
+DelimFileAdapter<double>::writeElem(std::ostream& stream,
+                                    const double& elem,
+                                    const unsigned& prec) const {
+    stream << std::setprecision(prec) << elem;
+}
+
+template<>
+inline void 
+DelimFileAdapter<SimTK::Vec2>::writeElem(std::ostream& stream,
+                                         const SimTK::Vec2& elem,
+                                         const unsigned& prec) const {
+    stream                    << std::setprecision(prec) << elem[0]
+           << _compDelimWrite << std::setprecision(prec) << elem[1];
+}
+
+template<>
+inline void 
+DelimFileAdapter<SimTK::Vec3>::writeElem(std::ostream& stream,
+                                         const SimTK::Vec3& elem,
+                                         const unsigned& prec) const {
+    stream                    << std::setprecision(prec) << elem[0]
+           << _compDelimWrite << std::setprecision(prec) << elem[1]
+           << _compDelimWrite << std::setprecision(prec) << elem[2];
+}
+
+template<>
+inline void 
+DelimFileAdapter<SimTK::Vec4>::writeElem(std::ostream& stream,
+                                         const SimTK::Vec4& elem,
+                                         const unsigned& prec) const {
+    stream << std::setprecision(prec) << elem[0];
+    for(auto i = 1u; i < 4; ++i)
```

Comment:
Again, because the initial goal for this PR was to add just 3 new types, I wrote *Full Specializations*. When you asked to support Quaternion and UnitVec, I just wrote more of them.